File: app/src/main/java/com/controlmoblie/asr/SenseVoiceModelManager.java
package com.controlmoblie.asr;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream;

import java.io.BufferedInputStream;
import java.io.File;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SenseVoiceModelManager {
    private static final String MODEL_URL = "https://github.com/k2-fsa/sherpa-onnx/releases/download/asr-models/icefall-asr-zipformer-streaming-wenetspeech-20230615.tar.bz2";
    private static final String MODEL_DIR_NAME = "wenetspeech-zipformer";
    private static final String TEMP_FILE_NAME = "asr-model.tar.bz2.tmp";
    private static final String TAG = "SenseVoiceModel";

    private static final Handler mainHandler = new Handler(Looper.getMainLooper());
    private static final ExecutorService executor = Executors.newSingleThreadExecutor();

    public interface DownloadProgressCallback {
        void onProgress(float progress);
    }

    public static boolean isModelReady(Context context) {
        File modelDir = new File(context.getFilesDir(), MODEL_DIR_NAME);
        if (!modelDir.exists() || !modelDir.isDirectory()) return false;
        List<File> onnxFiles = findRecursive(modelDir, name -> name.endsWith(".onnx"));
        return !onnxFiles.isEmpty();
    }

    public static String getModelDir(Context context) {
        return new File(context.getFilesDir(), MODEL_DIR_NAME).getAbsolutePath();
    }

    public static String getModelPath(Context context) {
        return getModelDir(context);
    }

    private static List<File> findRecursive(File dir, java.util.function.Predicate<String> predicate) {
        List<File> result = new ArrayList<>();
        collectFiles(dir, predicate, result);
        return result;
    }

    private static void collectFiles(File dir, java.util.function.Predicate<String> predicate, List<File> result) {
        File[] files = dir.listFiles();
        if (files == null) return;
        for (File f : files) {
            if (f.isDirectory()) {
                collectFiles(f, predicate, result);
            } else if (predicate.test(f.getName())) {
                result.add(f);
            }
        }
    }

    public static String[] getEncoderDecoderJoiner(Context context) {
        File dir = new File(context.getFilesDir(), MODEL_DIR_NAME);
        List<File> files = findRecursive(dir, name -> name.endsWith(".onnx"));
        if (files.isEmpty()) return null;
        String encoder = null, decoder = null, joiner = null;
        for (File f : files) {
            if (f.getName().toLowerCase().contains("encoder")) encoder = f.getAbsolutePath();
            else if (f.getName().toLowerCase().contains("decoder")) decoder = f.getAbsolutePath();
            else if (f.getName().toLowerCase().contains("joiner")) joiner = f.getAbsolutePath();
        }
        if (encoder != null && decoder != null && joiner != null) {
            return new String[]{encoder, decoder, joiner};
        }
        String single = files.get(0).getAbsolutePath();
        return new String[]{single, single, single};
    }

    public static String findTokensFile(Context context) {
        File dir = new File(context.getFilesDir(), MODEL_DIR_NAME);
        List<File> files = findRecursive(dir, name -> name.equals("tokens.txt"));
        return files.isEmpty() ? null : files.get(0).getAbsolutePath();
    }

    public static void downloadAndExtract(Context context, DownloadProgressCallback callback) {
        executor.execute(() -> {
            File modelDir = new File(context.getFilesDir(), MODEL_DIR_NAME);
            if (isModelReady(context)) {
                mainHandler.post(() -> callback.onProgress(1f));
                return;
            }

            File tmpFile = new File(context.getCacheDir(), TEMP_FILE_NAME);
            try {
                mainHandler.post(() -> callback.onProgress(0f));
                URL url = new URL(MODEL_URL);
                java.net.HttpURLConnection connection = (java.net.HttpURLConnection) url.openConnection();
                connection.setInstanceFollowRedirects(true);
                connection.setConnectTimeout(60000);
                connection.setReadTimeout(120000);
                connection.connect();
                int responseCode = connection.getResponseCode();
                Log.d(TAG, "Download response: code=" + responseCode + ", contentLength=" + connection.getContentLength());
                if (responseCode != 200) {
                    Log.e(TAG, "Download failed with HTTP " + responseCode);
                    return;
                }
                long fileLength = connection.getContentLengthLong();
                java.io.InputStream input = connection.getInputStream();
                java.io.FileOutputStream output = new java.io.FileOutputStream(tmpFile);
                byte[] buffer = new byte[8192];
                long totalRead = 0;
                int bytesRead;
                while ((bytesRead = input.read(buffer)) != -1) {
                    output.write(buffer, 0, bytesRead);
                    totalRead += bytesRead;
                    if (fileLength > 0) {
                        final float progress = totalRead / (float) fileLength * 0.7f;
                        mainHandler.post(() -> callback.onProgress(progress));
                    }
                }
                output.close();
                input.close();

                Log.d(TAG, "Download complete: " + tmpFile.length() + " bytes");
                if (tmpFile.length() < 1000) {
                    Log.e(TAG, "Downloaded file too small, likely not a valid model");
                    return;
                }

                mainHandler.post(() -> callback.onProgress(0.75f));
                if (modelDir.exists()) deleteDir(modelDir);
                modelDir.mkdirs();

                Log.d(TAG, "Starting extraction...");
                extractTarBz2(tmpFile, modelDir, fraction -> {
                    mainHandler.post(() -> callback.onProgress(0.75f + fraction * 0.25f));
                });
                Log.d(TAG, "Extraction complete");

                mainHandler.post(() -> callback.onProgress(1f));
            } catch (Exception e) {
                Log.e(TAG, "Download/extract failed", e);
            } finally {
                tmpFile.delete();
            }
        });
    }

    private static void deleteDir(File dir) {
        File[] files = dir.listFiles();
        if (files != null) {
            for (File f : files) {
                if (f.isDirectory()) deleteDir(f);
                else f.delete();
            }
        }
        dir.delete();
    }

    private static void extractTarBz2(File tarBz2File, File destDir, java.util.function.Consumer<Float> onProgress) {
        int entryCount = 0;
        try (BufferedInputStream bis = new BufferedInputStream(new java.io.FileInputStream(tarBz2File));
             BZip2CompressorInputStream bzIn = new BZip2CompressorInputStream(bis);
             TarArchiveInputStream tarIn = new TarArchiveInputStream(bzIn)) {
            org.apache.commons.compress.archivers.tar.TarArchiveEntry entry;
            while ((entry = tarIn.getNextTarEntry()) != null) {
                if (!entry.isDirectory() && entry.getName().indexOf('/') >= 0) {
                    String relativePath = entry.getName().substring(entry.getName().indexOf('/') + 1);
                    if (!relativePath.isEmpty()) entryCount++;
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to count entries", e);
        }

        Log.d(TAG, "Total entries to extract: " + entryCount);
        int total = Math.max(entryCount, 1);
        final int[] extracted = {0};
        try (BufferedInputStream bis = new BufferedInputStream(new java.io.FileInputStream(tarBz2File));
             BZip2CompressorInputStream bzIn = new BZip2CompressorInputStream(bis);
             TarArchiveInputStream tarIn = new TarArchiveInputStream(bzIn)) {
            org.apache.commons.compress.archivers.tar.TarArchiveEntry entry;
            while ((entry = tarIn.getNextTarEntry()) != null) {
                if (!entry.isDirectory()) {
                    String entryName = entry.getName();
                    int slashIdx = entryName.indexOf('/');
                    String relativePath = slashIdx >= 0 ? entryName.substring(slashIdx + 1) : entryName;
                    if (!relativePath.isEmpty()) {
                        File destFile = new File(destDir, relativePath);
                        File parent = destFile.getParentFile();
                        if (parent != null) parent.mkdirs();
                        try (java.io.FileOutputStream out = new java.io.FileOutputStream(destFile)) {
                            byte[] buf = new byte[8192];
                            int len;
                            while ((len = tarIn.read(buf)) != -1) {
                                out.write(buf, 0, len);
                            }
                        }
                        extracted[0]++;
                        onProgress.accept(extracted[0] / (float) total);
                        Log.d(TAG, "Extracted: " + relativePath + " (" + destFile.length() + " bytes)");
                    }
                }
            }
            Log.d(TAG, "Extracted " + extracted[0] + " files to " + destDir);
        } catch (Exception e) {
            Log.e(TAG, "Extraction failed", e);
        }
    }
}

File: app/src/main/java/com/controlmoblie/asr/SenseVoiceModelManager.kt
package com.controlmoblie.asr

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream
import java.io.BufferedInputStream
import java.io.File
import java.net.URL

object SenseVoiceModelManager {

    private const val MODEL_URL = "https://github.com/k2-fsa/sherpa-onnx/releases/download/asr-models/icefall-asr-zipformer-streaming-wenetspeech-20230615.tar.bz2"
    private const val MODEL_DIR_NAME = "wenetspeech-zipformer"
    private const val TEMP_FILE_NAME = "asr-model.tar.bz2.tmp"

    private const val TAG = "SenseVoiceModel"
    private val mainHandler = Handler(Looper.getMainLooper())

    fun isModelReady(context: Context): Boolean {
        val modelDir = File(context.filesDir, MODEL_DIR_NAME)
        if (!modelDir.exists() || !modelDir.isDirectory) return false
        val onnxFiles = modelDir.listFiles { f -> f.name.endsWith(".onnx") }
        return onnxFiles != null && onnxFiles.isNotEmpty()
    }

    fun getModelDir(context: Context): String {
        return File(context.filesDir, MODEL_DIR_NAME).absolutePath
    }

    fun getModelPath(context: Context): String = getModelDir(context)

    fun getEncoderDecoderJoiner(context: Context): Triple<String, String, String>? {
        val dir = File(context.filesDir, MODEL_DIR_NAME)
        val files = dir.listFiles { f -> f.name.endsWith(".onnx") } ?: return null
        val encoder = files.find { it.name.contains("encoder", ignoreCase = true) }?.absolutePath
        val decoder = files.find { it.name.contains("decoder", ignoreCase = true) }?.absolutePath
        val joiner = files.find { it.name.contains("joiner", ignoreCase = true) }?.absolutePath
        if (encoder != null && decoder != null && joiner != null) {
            return Triple(encoder, decoder, joiner)
        }
        val single = files.firstOrNull()
        return if (single != null) Triple(single.absolutePath, single.absolutePath, single.absolutePath) else null
    }

    suspend fun downloadAndExtract(context: Context, onProgress: (Float) -> Unit): Boolean {
        return withContext(Dispatchers.IO) {
            val modelDir = File(context.filesDir, MODEL_DIR_NAME)
            if (isModelReady(context)) {
                withContext(Dispatchers.Main) { onProgress(1f) }
                return@withContext true
            }

            val tmpFile = File(context.cacheDir, TEMP_FILE_NAME)
            try {
                withContext(Dispatchers.Main) { onProgress(0f) }
                val url = URL(MODEL_URL)
                val connection = (url.openConnection() as java.net.HttpURLConnection).apply {
                    instanceFollowRedirects = true
                    connectTimeout = 60000
                    readTimeout = 120000
                }
                connection.connect()
                val responseCode = connection.responseCode
                Log.d(TAG, "Download response: code=$responseCode, contentLength=${connection.contentLength}")
                if (responseCode != 200) {
                    Log.e(TAG, "Download failed with HTTP $responseCode")
                    return@withContext false
                }
                val fileLength = connection.contentLengthLong
                connection.getInputStream().use { input ->
                    tmpFile.outputStream().use { output ->
                        val buffer = ByteArray(8192)
                        var totalRead = 0L
                        var bytesRead: Int
                        while (input.read(buffer).also { bytesRead = it } != -1) {
                            output.write(buffer, 0, bytesRead)
                            totalRead += bytesRead
                            if (fileLength > 0) {
                                withContext(Dispatchers.Main) {
                                    onProgress(totalRead.toFloat() / fileLength * 0.7f)
                                }
                            }
                        }
                    }
                }

                Log.d(TAG, "Download complete: ${tmpFile.length()} bytes")
                if (tmpFile.length() < 1000) {
                    Log.e(TAG, "Downloaded file too small, likely not a valid model")
                    return@withContext false
                }

                withContext(Dispatchers.Main) { onProgress(0.75f) }
                if (modelDir.exists()) modelDir.deleteRecursively()
                modelDir.mkdirs()

                Log.d(TAG, "Starting extraction...")
                extractTarBz2(tmpFile, modelDir) { fraction ->
                    mainHandler.post { onProgress(0.75f + fraction * 0.25f) }
                }
                Log.d(TAG, "Extraction complete")

                withContext(Dispatchers.Main) { onProgress(1f) }
                isModelReady(context)
            } catch (e: Exception) {
                e.printStackTrace()
                false
            } finally {
                tmpFile.delete()
            }
        }
    }

    private fun extractTarBz2(tarBz2File: File, destDir: File, onProgress: (Float) -> Unit) {
        var entryCount = 0
        BufferedInputStream(tarBz2File.inputStream()).use { bis ->
            BZip2CompressorInputStream(bis).use { bzIn ->
                TarArchiveInputStream(bzIn).use { tarIn ->
                    var entry = tarIn.nextTarEntry
                    while (entry != null) {
                        if (!entry.isDirectory && entry.name.indexOf('/') >= 0) {
                            val relativePath = entry.name.substring(entry.name.indexOf('/') + 1)
                            if (relativePath.isNotBlank()) entryCount++
                        }
                        entry = tarIn.nextTarEntry
                    }
                }
            }
        }
        Log.d(TAG, "Total entries to extract: $entryCount")

        val total = maxOf(entryCount, 1)
        var extracted = 0
        BufferedInputStream(tarBz2File.inputStream()).use { bis ->
            BZip2CompressorInputStream(bis).use { bzIn ->
                TarArchiveInputStream(bzIn).use { tarIn ->
                    var entry = tarIn.nextTarEntry
                    while (entry != null) {
                        if (!entry.isDirectory) {
                            val entryName = entry.name
                            val slashIdx = entryName.indexOf('/')
                            val relativePath = if (slashIdx >= 0) entryName.substring(slashIdx + 1) else entryName
                            if (relativePath.isNotBlank()) {
                                val destFile = File(destDir, relativePath)
                                destFile.parentFile?.mkdirs()
                                destFile.outputStream().use { out ->
                                    tarIn.copyTo(out)
                                }
                                extracted++
                                onProgress(extracted.toFloat() / total)
                                Log.d(TAG, "Extracted: $relativePath (${destFile.length()} bytes)")
                            }
                        }
                        entry = tarIn.nextTarEntry
                    }
                    Log.d(TAG, "Extracted $extracted files to $destDir")
                }
            }
        }
    }
}
